### PR TITLE
feat(protocol-designer): transfer tools advanced settings and batch edit transfer

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-870, partially addresses AUTH-926

# Overview

This wires up the advanced settings for the transfer tools and the batch edit transfer tools. The modals still need to be updated though: https://opentrons.atlassian.net/browse/AUTH-928

Screenshot 2024-10-15 at 21 44 11
Screenshot 2024-10-15 at 21 44 27

## Test Plan and Hands on Testing

Create a protocol and add a transfer steps. look through the tools and make sure the advanced settings works as expected

additionally, test out a multi-dispense (distribute) step where you aspirate 1 well and dispense into more than 1 well, checkout the distribute options in the dispense advanced settings and make sure it works as expected

finally, test out the move liquid batch edit toolbox (command + click on 2 transfer steps to access it)

Note that the modals are not updated yet and most of them are using the old code. there is a followup ticket for those.

## Changelog

- add a bunch new components for the different advanced settings fields in transfer tools. they will be used in the mix tools and batch edit tools as well i believe
- wires up the new components to be used in transfer tools
- wires up the move liquid batch edit tools

## Risk assessment

low

### CODE REVIEW

**Review by greatheart2 — APPROVED** [review]

File: protocol-designer/src/pages/Designer/ProtocolSteps/BatchEditToolbox/BatchEditMoveLiquidTools.tsx [hunk 1]
```diff
@@ -1,3 +1,251 @@
...
+  }
+  const getWellOrderFieldValue = (
+    name: string
+  ): WellOrderOption | null | undefined => {
+    const val = propsForFields[name]?.value
+    if (val === 'l2r' || val === 'r2l' || val === 't2b' || val === 'b2t') {
+      return val
+    } else {
+      return null
+    }
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/BatchEditToolbox/BatchEditMoveLiquidTools.tsx:72]: nit

```ts
const validValues = ['l2r', 'r2l', 't2b', 'b2t'];
return validValues.includes(val) ? val : null;
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/BatchEditToolbox/index.tsx [hunk 2]
```diff
@@ -10,15 +10,17 @@ import {
...
 } from '../../../../step-forms/actions'
+import { maskField } from '../../../../steplist/fieldLevel'
+import { getBatchEditFormHasUnsavedChanges } from '../../../../step-forms/selectors'
+import { makeBatchEditFieldProps } from './utils'
 import { BatchEditMoveLiquidTools } from './BatchEditMoveLiquidTools'
 import { BatchEditMixTools } from './BatchEditMixTools'
-// import { maskField } from '../../../../steplist/fieldLevel'
 
-// import type { StepFieldName } from '../../../../steplist/fieldLevel'
+import type { StepFieldName } from '../../../../steplist/fieldLevel'
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/BatchEditToolbox/index.tsx]:
```suggestion
import type { ThunkDispatch } from 'redux-thunk'
import type { StepFieldName } from '../../../../steplist/fieldLevel'
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/BlowoutLocationField.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import type { Options } from '@opentrons/components'
+import type { FieldProps } from '../types'
+
+type BlowoutLocationDropdownProps = FieldProps & {
+  options: Options
+}
+
+export const BlowoutLocationField = (
+  props: BlowoutLocationDropdownProps
+): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/BlowoutLocationField.tsx]:
```suggestion
export function BlowoutLocationField(
  props: BlowoutLocationDropdownProps
): JSX.Element {
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/DisposalField.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  pipette: string | null
+  propsForFields: FieldPropsByName
+  stepType: StepType
+  volume: string | null
+  aspirate_airGap_checkbox?: boolean | null
+  aspirate_airGap_volume?: string | null
+  tipRack?: string | null
+}
+
+export const DisposalField = (props: DisposalFieldProps): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/DisposalField.tsx]:
```suggestion
export function DisposalField(props: DisposalFieldProps): JSX.Element {
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+interface InitialState {
+  isPristine: boolean
+  modalUseDefault: boolean
+  showModal: boolean
+  modalFlowRate?: string | null
+}
+
+export const FlowRateInput = (props: FlowRateInputProps): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]:
```suggestion
export function FlowRateInput(props: FlowRateInputProps): JSX.Element {
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 6]
```diff
@@ -0,0 +1,254 @@
+import * as React from 'react'
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]:
```suggestion
import { useState } from 'react'
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  ])
+
+  const initialState: InitialState = {
+    isPristine: true,
+    modalFlowRate: props.value ? String(props.value) : null,
+    modalUseDefault: !props.value && !isIndeterminate,
+    showModal: false,
+  }
+
+  const [isPristine, setIsPristine] = React.useState<
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]:
```suggestion
  const [isPristine, setIsPristine] = useState<
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    InitialState['isPristine']
+  >(initialState.isPristine)
+
+  const [modalFlowRate, setModalFlowRate] = React.useState<
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]: same as the above

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    InitialState['modalFlowRate']
+  >(initialState.modalFlowRate)
+
+  const [modalUseDefault, setModalUseDefault] = React.useState<
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]: same as the above

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    InitialState['modalUseDefault']
+  >(initialState.modalUseDefault)
+
+  const [showModal, setShowModal] = React.useState<InitialState['showModal']>(
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]: same as the above

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+  Flex,
+  useHoverTooltip,
+  InputField,
+  Modal,
+  SecondaryButton,
+  PrimaryButton,
+  Tooltip,
+} from '@opentrons/components'
+import { getMainPagePortalEl } from '../../../../../components/portals/MainPageModalPortal'
+import type { FieldProps } from '../types'
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx:17]:
```suggestion
import type { ChangeEvent } from 'react'
import type { FieldProps } from '../types'
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+    setIsPristine(false)
+
+    if (allowSave) {
+      const newFlowRate = modalUseDefault ? null : Number(modalFlowRate)
+      props.updateValue(newFlowRate)
+      resetModalState()
+    }
+  }
+
+  const handleChangeRadio = (e: React.ChangeEvent<HTMLSelectElement>): void => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]:
```suggestion
  const handleChangeRadio = (e: ChangeEvent<HTMLSelectElement>): void => {
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    setModalUseDefault(e.target.value !== 'custom')
+  }
+
+  const handleChangeNumber = (e: React.ChangeEvent<HTMLInputElement>): void => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx]: same as the above

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+      isIndeterminate={isIndeterminate && modalFlowRate === null}
+      name={`${name}_customFlowRate`}
+      onChange={handleChangeNumber}
+      units={t('application:units.microliterPerSec')}
+      value={`${modalFlowRate || ''}`}
+    />
+  )
+
+  //    TODO: update the modal
+  const FlowRateModal =
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx:165]: plan to export this as `FlowRateModal`?

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/PositionField.tsx [hunk 15]
```diff
@@ -0,0 +1,212 @@
+import * as React from 'react'
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/PositionField.tsx]:
```suggestion
import { useState } from 'react'
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/PositionField.tsx [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+    value: rawZValue,
+    updateValue: zUpdateValue,
+    tooltipContent,
+    isIndeterminate,
+    disabled,
+  } = propsForFields[zField]
+
+  const { t, i18n } = useTranslation(['application', 'protocol_steps'])
+  const [targetProps, tooltipProps] = useHoverTooltip()
+  const [isModalOpen, setModalOpen] = React.useState<boolean>(false)
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/PositionField.tsx]:
```suggestion
  const [isModalOpen, setModalOpen] = useState<boolean>(false)
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/WellsOrderField.tsx [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  prefix: 'aspirate' | 'dispense' | 'mix'
+  firstName: string
+  secondName: string
+  updateFirstWellOrder: FieldProps['updateValue']
+  updateSecondWellOrder: FieldProps['updateValue']
+  firstValue?: WellOrderOption | null
+  secondValue?: WellOrderOption | null
+}
+
+export const WellsOrderField = (props: WellsOrderFieldProps): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/WellsOrderField.tsx]:
```suggestion
export function WellsOrderField(props: WellsOrderFieldProps): JSX.Element {
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx (hunk 14 above)
Comment (jerader) [reply protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/PipetteFields/FlowRateInput.tsx:165]: yeah i think that can be done in a followup

**Comment by ncdiehl11** [comment]
Everything is looking good but looks like you can save the form with mix being checked with empty values for repetitions and volume
Screenshot 2024-10-17 at 11 30 32 AM

**Review by koji — APPROVED** [review]
lgtm 🤖 
It’s up to you how to handle Nick’s feedback. this pr or a following pr

**Comment by jerader** [comment]
> Everything is looking good but looks like you can save the form with mix being checked with empty values for repetitions and volume Screenshot 2024-10-17 at 11 30 32 AM

@ncdiehl11 oof good catch, i think this is old PD behavior. I'll make a follow-up ticket to address it!

**Comment by jerader** [comment]
the followup ticket: https://opentrons.atlassian.net/browse/AUTH-961

**Pull request merged**